controlplane/pkg/nsm/nsm.go

```diff
@@ -0,0 +1,411 @@
+	} else {
+		message := srv.createRemoteNSMRequest(endpoint, requestConnection, dataplane)
+		return message
+	}
+}
+
+func (srv *networkServiceManager) isLocalEndpoint(endpoint *registry.NSERegistration) bool {
+	return srv.getNetworkServiceManagerName() == endpoint.GetNetworkServiceManager().GetName()
+}
+
+func (srv *networkServiceManager) createNSEClient(endpoint *registry.NSERegistration) (nsm.NetworkServiceClient, error) {
+	if srv.isLocalEndpoint(endpoint) {
+		client, conn, err := srv.serviceRegistry.EndpointConnection(endpoint)
+		if err != nil {
+			return nil, err
+		}
+		return &endpointClient{connection: conn, client: client}, nil
+	} else {
+		client, conn, err := srv.serviceRegistry.RemoteNetworkServiceClient(endpoint.GetNetworkServiceManager())
+		if err != nil {
+			return nil, err
+		}
+		return &nsmClient{client: client, connection: conn}, nil
+	}
+}
+
+func (srv *networkServiceManager) performNSERequest(request nsm.NSMRequest, endpoint *registry.NSERegistration, ctx context.Context, requestConnection nsm.NSMConnection, dp *model.Dataplane) (*model.ClientConnection, error) {
+	client, err := srv.createNSEClient(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	defer client.Cleanup()
+
+	message := srv.CreateNSERequest(endpoint, requestConnection, dp)
+	nseConnection, e := client.Request(ctx, message)
+
+	if e != nil {
+		logrus.Errorf("error requesting networkservice from %+v with message %#v error: %s", endpoint, message, e)
+		return nil, e
+	}
+
+	err = srv.validateNSEConnection(nseConnection)
+	if err != nil {
+		return nil, err
+	}
+
+	err = requestConnection.UpdateContext(nseConnection.GetContext())
+	if err != nil {
+		err = fmt.Errorf("failure Validating NSE Connection: %s", err)
+		return nil, err
+	}
+	srv.updateConnectionParameters(nseConnection, endpoint)
+
+	dpApiConnection := srv.createCrossConnect(requestConnection, endpoint, request, nseConnection)
+	clientConnection := &model.ClientConnection{
+		ConnectionId: requestConnection.GetId(),
+		Xcon:         dpApiConnection,
+		Endpoint:     endpoint,
+		Dataplane:    dp,
+	}
+	return clientConnection, nil
+}
+
+func (srv *networkServiceManager) createCrossConnect(requestConnection nsm.NSMConnection, endpoint *registry.NSERegistration, request nsm.NSMRequest, nseConnection nsm.NSMConnection) *crossconnect.CrossConnect {
+	dpApiConnection := &crossconnect.CrossConnect{
+		Id:      requestConnection.GetId(),
+		Payload: endpoint.GetNetworkService().GetPayload(),
+	}
+
+	// We handling request from remote NSM
+	if request.IsRemote() {
+		dpApiConnection.Source = &crossconnect.CrossConnect_RemoteSource{
+			RemoteSource: requestConnection.(*remote_connection.Connection),
+		}
+	} else {
+		dpApiConnection.Source = &crossconnect.CrossConnect_LocalSource{
+			LocalSource: requestConnection.(*connection.Connection),
+		}
+	}
+
+	// We handling request from local or remote endpoint.
+	//TODO: in case of remote NSE( different cluster case, this method should be changed)
```

Comment:
That's a nice observation. Are we saying we'll have to implement 3 NSE types: local, remote, external?

Reply:
I'm not sure how it will be, this is one of possible scenarios.